Using C library fails to compile with 2.2.0 but works with 2.1.0

Compiling using 2.2.0 gives me this error:

```
In file included from ./local/include/treelite/c_api.h:15,
                 from mm-xgboost-codegen.c:8:
./local/include/treelite/c_api_common.h:54:8: error: expected identifier or ‘(’ before string constant                                                                 
   54 | extern "C" {
      |        ^~~
```

From a look at the changes in 2.2.0, I think the issue is here: https://github.com/dmlc/treelite/blob/1f9c5054ad7433fa88623fccf0ae46a6ba6a27c6/include/treelite/c_api_common.h#L54 from https://github.com/dmlc/treelite/pull/325.

I think instead of `extern "C"` we'd want to use the macro `TREELITE_EXTERN_C`. Or maybe `TREELITE_DLL`, I'm not sure.

Thank you!

--- a/include/treelite/c_api_common.h
+++ b/include/treelite/c_api_common.h
@@ -51,9 +51,13 @@ TREELITE_DLL int TreeliteRegisterLogCallback(void (*callback)(const char*));
  */
 TREELITE_DLL const char* TreeliteQueryTreeliteVersion(void);
 
+#ifdef __cplusplus
 extern "C" {
   extern const char* TREELITE_VERSION;
 }
+#else
+extern const char* TREELITE_VERSION;
+#endif
 
 /*!
  * \defgroup dmatrix

--- a/tests/example_app/CMakeLists.txt
+++ b/tests/example_app/CMakeLists.txt
@@ -13,7 +13,15 @@ find_package(Treelite REQUIRED)
 add_executable(cpp_example example.cc)
 target_link_libraries(cpp_example PRIVATE treelite::treelite treelite::treelite_runtime)
 
+add_executable(c_example example.c)
+target_link_libraries(c_example PRIVATE treelite::treelite treelite::treelite_runtime)
+
 set_target_properties(cpp_example PROPERTIES
     CXX_STANDARD 14
     CXX_STANDARD_REQUIRED YES
+)
+
+set_target_properties(c_example PROPERTIES
+    C_STANDARD 99
+    C_STANDARD_REQUIRED YES
 )

--- a/tests/example_app/CMakeLists.txt
+++ b/tests/example_app/CMakeLists.txt
@@ -10,10 +10,10 @@ endif()
 
 find_package(Treelite REQUIRED)
 
-add_executable(example example.cc)
-target_link_libraries(example PRIVATE treelite::treelite_static treelite::treelite_runtime_static)
+add_executable(cpp_example example.cc)
+target_link_libraries(cpp_example PRIVATE treelite::treelite treelite::treelite_runtime)
 
-set_target_properties(example PROPERTIES
+set_target_properties(cpp_example PROPERTIES
     CXX_STANDARD 14
     CXX_STANDARD_REQUIRED YES
 )

--- a/tests/example_app/example.c
+++ b/tests/example_app/example.c
@@ -64,6 +64,8 @@ int main() {
   output = (float*)malloc(output_alloc_size * sizeof(float));
   safe_treelite(TreeliteGTILPredict(model, input, num_row, output, 0, &out_result_size));
 
+  printf("TREELITE_VERSION = %s\n", TREELITE_VERSION);
+
   for (size_t i = 0; i < num_row; ++i) {
     printf("Input %d: [%f", (int)i, input[i * num_col]);
     for (size_t j = 1; j < num_col; ++j) {

--- a/tests/example_app/example.c
+++ b/tests/example_app/example.c
@@ -1,0 +1,78 @@
+#include <treelite/c_api.h>
+#include <treelite/c_api_runtime.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stddef.h>
+
+#define safe_treelite(call) {  \
+  int err = (call);  \
+  if (err == -1) {  \
+    fprintf(stderr, "%s:%d: error in %s: %s\n", __FILE__, __LINE__, #call,  \
+            TreeliteGetLastError());  \
+    exit(1);  \
+  }  \
+}
+
+ModelHandle BuildModel() {
+  ModelHandle model;
+  ModelBuilderHandle builder;
+  TreeBuilderHandle tree;
+  float threshold_value = 0.0f;
+  float leaf0_value = -1.0f;
+  float leaf1_value = 1.0f;
+  ValueHandle threshold, leaf0, leaf1;
+  safe_treelite(TreeliteTreeBuilderCreateValue(&threshold_value, "float32", &threshold));
+  safe_treelite(TreeliteTreeBuilderCreateValue(&leaf0_value, "float32", &leaf0));
+  safe_treelite(TreeliteTreeBuilderCreateValue(&leaf1_value, "float32", &leaf1));
+
+  safe_treelite(TreeliteCreateTreeBuilder("float32", "float32", &tree));
+  safe_treelite(TreeliteTreeBuilderCreateNode(tree, 0));
+  safe_treelite(TreeliteTreeBuilderCreateNode(tree, 1));
+  safe_treelite(TreeliteTreeBuilderCreateNode(tree, 2));
+  safe_treelite(TreeliteTreeBuilderSetNumericalTestNode(tree, 0, 0, "<", threshold, 1, 1, 2));
+  safe_treelite(TreeliteTreeBuilderSetLeafNode(tree, 1, leaf0));
+  safe_treelite(TreeliteTreeBuilderSetLeafNode(tree, 2, leaf1));
+  safe_treelite(TreeliteTreeBuilderSetRootNode(tree, 0));
+
+  safe_treelite(TreeliteCreateModelBuilder(2, 1, 0, "float32", "float32", &builder));
+  safe_treelite(TreeliteModelBuilderInsertTree(builder, tree, -1));
+  safe_treelite(TreeliteModelBuilderCommitModel(builder, &model));
+
+  // Clean up
+  safe_treelite(TreeliteTreeBuilderDeleteValue(threshold));
+  safe_treelite(TreeliteTreeBuilderDeleteValue(leaf0));
+  safe_treelite(TreeliteTreeBuilderDeleteValue(leaf1));
+  safe_treelite(TreeliteDeleteTreeBuilder(tree));
+  safe_treelite(TreeliteDeleteModelBuilder(builder));
+
+  return model;
+}
+
+int main() {
+  ModelHandle model = BuildModel();
+  size_t num_row = 5;
+  size_t num_col = 2;
+  size_t output_alloc_size;
+  float input[10] = {-2.0f, 0.0f,
+                     -1.0f, 0.0f,
+                      0.0f, 0.0f,
+                      1.0f, 0.0f,
+                      2.0f, 0.0f};
+  float* output;
+  size_t out_result_size;
+  safe_treelite(TreeliteGTILGetPredictOutputSize(model, num_row, &output_alloc_size));
+  output = (float*)malloc(output_alloc_size * sizeof(float));
+  safe_treelite(TreeliteGTILPredict(model, input, num_row, output, 0, &out_result_size));
+
+  for (size_t i = 0; i < num_row; ++i) {
+    printf("Input %d: [%f", (int)i, input[i * num_col]);
+    for (size_t j = 1; j < num_col; ++j) {
+      printf(", %f", input[i * num_col + j]);
+    }
+    printf("], output: %f\n", output[i]);
+  }
+
+  free(output);
+
+  return 0;
+}

--- a/tests/example_app/example.cc
+++ b/tests/example_app/example.cc
@@ -1,6 +1,7 @@
 #include <treelite/tree.h>
 #include <treelite/frontend.h>
 #include <treelite/compiler.h>
+#include <treelite/c_api_common.h>
 #include <iostream>
 #include <map>
 #include <memory>
@@ -11,6 +12,7 @@ using treelite::frontend::Value;
 using treelite::TypeInfo;
 
 int main(void) {
+  std::cout << "TREELITE_VERSION = " << TREELITE_VERSION << std::endl;
   auto tree = std::make_unique<TreeBuilder>(TypeInfo::kFloat32, TypeInfo::kFloat32);
   tree->CreateNode(0);
   tree->CreateNode(1);

--- a/tests/travis/run_test.sh
+++ b/tests/travis/run_test.sh
@@ -38,14 +38,14 @@ then
   conda activate python3
   conda --version
   python --version
-  conda install -c conda-forge ninja cmake
+  conda install -c conda-forge ninja cmake rapidjson fmt
 
   # Install Treelite C++ library into the Conda env
   set -x
   rm -rf build/
   mkdir build
   cd build
-  cmake .. -DCMAKE_INSTALL_PREFIX="$CONDA_PREFIX" -DCMAKE_INSTALL_LIBDIR="lib" -DBUILD_STATIC_LIBS=ON -GNinja
+  cmake .. -DCMAKE_INSTALL_PREFIX="$CONDA_PREFIX" -DCMAKE_INSTALL_LIBDIR="lib" -GNinja
   ninja install
 
   # Try compiling a sample application
@@ -55,7 +55,7 @@ then
   cd build
   cmake .. -GNinja
   ninja
-  ./example
+  ./cpp_example
 fi
 
 if [ ${TASK} == "python_test" ]

--- a/tests/travis/run_test.sh
+++ b/tests/travis/run_test.sh
@@ -56,6 +56,7 @@ then
   cmake .. -GNinja
   ninja
   ./cpp_example
+  ./c_example
 fi
 
 if [ ${TASK} == "python_test" ]

